From e6a6e4f1fbcb029bac0cb8eecdeb2879694e1ba8 Mon Sep 17 00:00:00 2001
From: Slatian <baschdel@disroot.org>
Date: Sat, 31 Aug 2024 12:14:45 +0200
Subject: [PATCH 1/2] Implement XDG_CONFIG_HOME resolution in shell to get rid
 qtpaths dependency.

Also used in other places of xdg-mime because the qtpaths replacement implements
the base-dir specification more correctly than previous shell hackery.

Fixes: #258
---
 scripts/xdg-mime.in         | 10 +++-------
 scripts/xdg-utils-common.in | 15 +++++++++++++++
 2 files changed, 18 insertions(+), 7 deletions(-)

diff --git a/scripts/xdg-mime.in b/scripts/xdg-mime.in
index 18056df..0af03f1 100644
--- a/scripts/xdg-mime.in
+++ b/scripts/xdg-mime.in
@@ -167,7 +167,7 @@ make_default_kde()
     vendor="$1"
     mimetype="$2"
     if [ "${KDE_SESSION_VERSION:-0}" -gt 4 ] ; then
-        default_dir="$(qtpaths --writable-path ConfigLocation)"
+        default_dir="$(get_xdg_config_home)"
         default_file="$default_dir/mimeapps.list"
     elif [ x"$KDE_SESSION_VERSION" = x"4" ]; then
         default_dir="$(kde4-config --path xdgdata-apps 2> /dev/null | cut -d ':' -f 1)"
@@ -294,10 +294,7 @@ make_default_generic()
     # $1 is vendor-name.desktop
     # $2 is mime/type
     # Add $2=$1 to XDG_CONFIG_HOME/mimeapps.list
-    #shellcheck disable=SC2153 # not misspelled
-    xdg_config_home="$XDG_CONFIG_HOME"
-    [ -n "$xdg_config_home" ] || xdg_config_home="$HOME/.config"
-    default_file="$xdg_config_home/mimeapps.list"
+    default_file="$(get_xdg_config_home)/mimeapps.list"
     if [ -L "$default_file" ]; then
         out_file="$(xdg_realpath "$default_file")"
     else
@@ -459,8 +456,7 @@ check_mimeapps_list()
 defapp_generic()
 {
     MIME="$1"
-    xdg_config_home="$XDG_CONFIG_HOME"
-    [ -n "$xdg_config_home" ] || xdg_config_home="$HOME/.config"
+    xdg_config_home="$(get_xdg_config_home)"
     #shellcheck disable=SC2153 # not misspelled
     xdg_config_dirs="$XDG_CONFIG_DIRS"
     [ -n "$xdg_config_dirs" ] || xdg_config_dirs="/etc/xdg"
diff --git a/scripts/xdg-utils-common.in b/scripts/xdg-utils-common.in
index 097322d..d9c0803 100644
--- a/scripts/xdg-utils-common.in
+++ b/scripts/xdg-utils-common.in
@@ -515,3 +515,18 @@ xdg_which()
 		return 1
 	fi
 }
+
+#----------------------------------------------------------------------------
+# Helper function that returns the value for `$XDG_CONFIG_HOME`,
+# independent of it being explicitly set or simply a fallback.
+
+# get_xdg_config_home
+get_xdg_config_home()
+{
+	# Only use XDG_CONFIG_HOME if it is an absolute path
+	if printf "%s" "$XDG_CONFIG_HOME" | grep -q '^/' ; then
+		printf "%s\n" "$XDG_CONFIG_HOME"
+	else
+		printf "%s\n" "$HOME/.config"
+	fi
+}
-- 
GitLab


From 23a1207c5826ecec0a5b0c5810b58e915e46258f Mon Sep 17 00:00:00 2001
From: Slatian <baschdel@disroot.org>
Date: Sat, 31 Aug 2024 12:39:28 +0200
Subject: [PATCH 2/2] Get rid of call to grep in get_xdg_config_home

---
 scripts/xdg-utils-common.in | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/scripts/xdg-utils-common.in b/scripts/xdg-utils-common.in
index d9c0803..7fae86b 100644
--- a/scripts/xdg-utils-common.in
+++ b/scripts/xdg-utils-common.in
@@ -524,9 +524,9 @@ xdg_which()
 get_xdg_config_home()
 {
 	# Only use XDG_CONFIG_HOME if it is an absolute path
-	if printf "%s" "$XDG_CONFIG_HOME" | grep -q '^/' ; then
-		printf "%s\n" "$XDG_CONFIG_HOME"
-	else
-		printf "%s\n" "$HOME/.config"
-	fi
+	# shellcheck disable=SC2153 # not misspelled
+	case "$XDG_CONFIG_HOME" in
+		/*) printf "%s\n" "$XDG_CONFIG_HOME" ;;
+		*) printf "%s\n" "$HOME/.config" ;;
+	esac
 }
-- 
GitLab

